src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/ChainedMethodCallWrapCheck.java

```diff
@@ -0,0 +1,496 @@
+     *
+     * @param maxCallsPerLine maximum number of method calls per lines if we are wrapping
+     *                        the method calls
+     */
+    public void setMaxCallsPerLine(int maxCallsPerLine) {
+        this.maxCallsPerLine = maxCallsPerLine;
+    }
+
+    /**
+     * Setter to specify maximum number of method calls if all chained method calls span over a
+     * single line. If chained method calls span over multiple lines then this property has no
+     * effect.
+     *
+     * @param maxSingleLineCalls maximum number of method calls if all chained method calls span
+     *                           over a single line
+     */
+    public void setMaxSingleLineCalls(int maxSingleLineCalls) {
+        this.maxSingleLineCalls = maxSingleLineCalls;
+    }
+
+    @Override
+    public int[] getDefaultTokens() {
+        return new int[] {
+            TokenTypes.METHOD_CALL,
+        };
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getDefaultTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return getAcceptableTokens();
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        if (isTopMostMethodCall(ast)
+                && matchesIdentifierPattern(ast)) {
+            if (isSingleLineMethodChain(ast)) {
+                checkMethodChainingSingleLine(ast);
+            }
+            else {
+                checkMethodChainingMultiLine(ast);
+            }
+        }
+    }
+
+    /**
+     * Whether {@code methodCallAst} is the top most method call in the
+     * abstract syntax tree.
+     *
+     * @param methodCallAst ast node of type {@link TokenTypes#METHOD_CALL}
+     * @return {@code true} if {@code methodCallAst} is the top most method call in the
+     *         abstract syntax tree
+     */
+    private static boolean isTopMostMethodCall(DetailAST methodCallAst) {
+        boolean result = false;
+        DetailAST ancestorAst = methodCallAst.getParent();
+        while (!TokenUtil.isOfType(ancestorAst, PARENT_OF_METHOD_CALL)) {
+            final DetailAST ancestorAstPrevSibling = ancestorAst.getPreviousSibling();
+            if (ancestorAstPrevSibling != null
+                    && !TokenUtil.isOfType(ancestorAst, TokenTypes.TYPECAST, TokenTypes.INDEX_OP)) {
+                result = true;
+                break;
+            }
+            ancestorAst = ancestorAst.getParent();
+        }
+        return result || TokenUtil.isOfType(ancestorAst, TokenTypes.EXPR, TokenTypes.LITERAL_CASE);
+    }
+
+    /**
+     * Whether the chained method calls match the
+     * {@link ChainedMethodCallWrapCheck#identifierPattern}.
+     *
+     * @param methodCallAst ast node of type {@link TokenTypes#METHOD_CALL}
+     * @return {@code true} if the chained method calls match the identifier pattern
+     */
+    private boolean matchesIdentifierPattern(DetailAST methodCallAst) {
+        final String getCompleteMethodChainName = FullIdent
```

Comment:
I think this would read better with `FullIdent` on next line with method calls.

Reply:
Done.

Reply:
Resolved.